src/node/stage/approved.rs

```diff
@@ -568,19 +570,61 @@ impl Approved {
         }
     }
 
-    pub fn handle_genesis_update(
+    pub fn handle_elders_update(&mut self, core: &mut Core, payload: EldersUpdate) -> Result<()> {
+        info!("Received {:?}", payload);
+
+        core.msg_filter.reset();
+
+        self.shared_state = SharedState::new(payload.elders_info);
+        self.section_keys_provider = SectionKeysProvider::new(None);
+        self.reset_parsec(core, payload.parsec_version)
+    }
+
+    pub fn handle_promote(
         &mut self,
         core: &mut Core,
-        genesis_prefix_info: GenesisPrefixInfo,
-        section_key: bls::PublicKey,
+        shared_state: SharedState,
+        parsec_version: u64,
     ) -> Result<()> {
-        info!("Received GenesisUpdate: {:?}", genesis_prefix_info);
+        if !shared_state.our_info().elders.contains_key(core.name()) {
+            debug!("ignore Promote - not actually promoted");
+            return Ok(());
+        }
 
         core.msg_filter.reset();
 
-        self.shared_state = SharedState::new(genesis_prefix_info.elders_info, section_key);
-        self.section_keys_provider = SectionKeysProvider::new(None);
-        self.reset_parsec(core, genesis_prefix_info.parsec_version)
+        // TODO: verify `shared_state` !
+        self.shared_state.update(shared_state)?;
+
+        self.reset_parsec(core, parsec_version)?;
+        self.gossip_timer_token = Some(core.timer.schedule(self.consensus_engine.gossip_period()));
+
+        info!("Promoted");
+
+        match self
+            .section_keys_provider
+            .finalise_dkg(core.name(), self.shared_state.our_info())
+        {
+            Ok(()) => (),
+            Err(RoutingError::InvalidElderDkgResult) => {
+                // Ignore `InvalidElderDkgResult` because it just means the DKG hasn't completed
+                // for us yet.
+                // TODO: check that we have an ongoing DKG instance corresponding to the latest
+                // EldersInfo.
+            }
+            Err(error) => return Err(error),
+        }
+
+        self.send_elders_update(core)?;
+        self.send_send_neighbour_info()?;
+        self.send_their_knowledge();
```

Comment:
I'm confused by this. This basically means we are voting to update the knowledge of our section by other sections to be up to date, but how do we know that to be the case?

Reply:
Same as above, this is mainly for the split case. Without this, there will be not-enough votes to consensus the knowledge update.

Reply:
OK, but how do we know that the knowledge is really up-to-date? Because what we are doing here is voting that every other section has up-to-date knowledge of out section, but I don't think that's necessarily the case. We can only know that if we receive a message from them and the message's `dst_key` field points to our latest key.

Reply:
The newly promoted nodes won't vote, so sometimes during split there will be not enough votes.
So here will have to vote during handle_promote (the updated shared_state shall contains the latest info)

Reply:
So the `send_their_knowledge` call should be replaced with a vote for `TheirKnowledeg` but only for the sibling section and only if there is a split.

Reply:
addressed by PR https://github.com/maidsafe/routing/pull/2170